File: longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left, right = 0, 0
        max_len = 0

        while right < len(s):
            if s[right] in s[left:right]:
                max_len = max(max_len, right - left)
                left = left + s[left:right].index(s[right]) + 1
            right += 1

        max_len = max(max_len, right - left)
        return max_len

File: longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {

        int len = s.length();
        if (len <= 1) {
            return len;
        }

        HashMap<Character, Integer> map = new HashMap<>();
        int start = 0;
        int maxLen = 0;

        for (int end = 0; end < len; end++) {
            char now = s.charAt(end);
            if (map.containsKey(now)) {
                start = Math.max(start, map.get(now) + 1);
            }
            map.put(now, end);
            maxLen = Math.max(maxLen, end - start + 1);
        }

        return maxLen;
    }
}